### PR TITLE
perf(lsp): explicitly invalidate script versions

### PR DESCRIPTION
**Pull request by nayeemrmn**
Alternative to #21542. Rather than clearing `scriptVersionCache` after each request, preserve it and invalidate entries on known file changes.

### CODE REVIEW

**Review by bartlomieju — APPROVED** [review]
Very cool! Let's wait on @dsherret's opinion before landing.

**Review by dsherret — APPROVED** [review]
LGTM

File: cli/tsc/99_main_compiler.js
Comment (dsherret) [inline cli/tsc/99_main_compiler.js:1038]: What happens when an https file is changed on the file system? For example, when using `{ "vendor": true }` in a deno.json then editing something in the vendor folder?

Comment (dsherret) [inline cli/tsc/99_main_compiler.js:1038]: By the way, this will cause problems where a script won't be invalidated when it changes on the file system via a different tool. This isn't what TypeScript does and I'm not sure we should do this (TypeScript creates a file watcher that invalidates when the file changes on the system for closed documents).

**Review by bartlomieju — CHANGES_REQUESTED** [review]
Discuss offline with David. Let's hold off landing this one before the v1.39 release.

**Comment by nayeemrmn** [comment]
We're reconsidering this because `op_script_version()` is not definitively a problem

**Pull request closed without merging**
